DESIGN.md

```diff
@@ -0,0 +1,111 @@
+
+#### SqlTable::Insert
+Insert will always insert the passed in tuple into the DataTable of the schema version number passed in
+
+#### SqlTable::Update
+- Update has three cases
+    1. The expected schema version matches the actual schema version
+        - The update will happen in place on the DataTable of the actual schema version
+    2. The expected schema version doesn’t match the actual schema version but the update doesn’t touch any columns not in the actual schema version
+        - The update will happen in place on the DataTable of the expected schema version
+    3. The expected schema version doesn’t match the actual schema version and the update touches not in the actual schema version. The following steps occur
+        - Retrieve the tuple from the actual version DataTable
+        - Transform the tuple to the expected version
+        - Delete the tuple from the actual version DataTable
+        - Insert the tuple into the expected version DataTable
+
+#### SqlTable::Scan
+- SqlTable will maintain its own slot iterator which will be used to iterate across all the schema version. The iterator interface exposed to the user will not change
+- The iterator for Scan must always begin on the latest version
+    - In cases where the user is interchanging scan call and update calls, if the scan iterator were to start on an older version then it is possible for a tuple retrieved from the scan to be updated, which could move it into the latest schema version
+    - This would mean that when the iterator gets to the latest schema version it will read all the tuples in the DataTable for that version and this would cause that tuple to have been read twice within the scan.
+
+#### SqlTable::Select
+- There are two cases
+    1. Expected schema version matches actual schema version
+        - The tuple is directly selected from the DataTable for that schema version
+    2. Expected schema version differs from the actual schema version
+        - The tuple is selected from the DataTable for that schema version and during this process it is transformed to the expected schema version
```

Comment:
Will the transformation also happen in the underlying storage? Or it is only a temporary result and then be thrown away. I think it is better to clarify that in the document.

Reply:
The transformation here is temporary.  The only time we persist the transformation is in case 3 for update where the tuple is moved between DataTables via a delete and insert.

Reply:
In the document, you said that your benchmark showed that reading a tuple which version doesn't match has a penalty to a factor of 5x. Consider the case that a hot tuple is in an older version. Each read needs to pay a 5x penalty. If the tuple is hot, we will transform the tuple again and again, and the overhead could be large. I think you may need some policy to migrate the tuple in such case, even we are only reading the tuple, to reduce the overhead.